roadmap/implementers-guide/src/runtime/router.md

```diff
@@ -27,9 +110,108 @@ No initialization routine runs for this module.
+  1. Checks that `processed_downward_messages` is at least 1 if `DownwardMessageQueues` for `recipient` is not empty.
+* `check_hrmp_watermark(P: ParaId, new_hrmp_watermark)`:
+  1. `new_hrmp_watermark` should be strictly greater than the value of `HrmpWatermarks` for `P` (if any).
+  1. `new_hrmp_watermark` must not be greater than the context's block number.
+  1. in ``HrmpChannelDigests`` for `P` an entry with the block number equal to `new_hrmp_watermark` should exist.
+* `verify_outbound_hrmp(sender: ParaId, Vec<OutboundHrmpMessage>)`:
+  1. For each horizontal message `M` with the channel `C` identified by `(sender, M.recipient)` check:
+      1. exists
+      1. `M`'s payload size summed with the `C.used_bytes` doesn't exceed a preconfigured limit `C.limit_used_bytes`.
+      1. `C.used_places + 1` doesn't exceed a preconfigured limit `C.limit_used_places`.
+
+Candidate Enactment:
+
+* `queue_outbound_hrmp(sender: ParaId, Vec<OutboundHrmpMessage>)`:
+  1. For each horizontal message `HM` with the channel `C` identified by `(sender, HM.recipient)`:
+    1. Append `HM` into `HrmpChannelContents` that corresponds to `C`.
+    1. Locate or create an entry in ``HrmpChannelDigests`` for `HM.recipient` and append `sender` into the entry's list.
+    1. Increment `C.used_places`
+    1. Increment `C.used_bytes` by `HM`'s payload size
+* `prune_hrmp(recipient, new_hrmp_watermark)`:
+  1. From ``HrmpChannelDigests`` for `recipient` remove all entries up to an entry with block number equal to `new_hrmp_watermark`.
+  1. From the removed digests construct a set of paras that sent new messages within the interval between the old and new watermarks.
+  1. For each channel `C` identified by `(sender, recipient)` for each `sender` coming from the set, prune messages up to the `new_hrmp_watermark`.
+  1. For each pruned message `M` from channel `C`:
+      1. Decrement `C.used_places`
+      1. Decrement `C.used_bytes` by `M`'s payload size.
+  1. Set `HrmpWatermarks` for `P` to be equal to `new_hrmp_watermark`
+* `prune_dmq(P: ParaId, processed_downward_messages)`:
+  1. Remove the first `processed_downward_messages` from the `DownwardMessageQueues` of `P`.
+
 * `queue_upward_messages(ParaId, Vec<UpwardMessage>)`:
   1. Updates `NeedsDispatch`, and enqueues upward messages into `RelayDispatchQueue` and modifies the respective entry in `RelayDispatchQueueSize`.
 
+The routines described below are for use within a session change and called by the `Paras` module.
+
+* `offboard_para(P: ParaId)`:
+  1. Remove all inbound channels of `P`, i.e. `(_, P)`,
+  1. Remove all outbound channels of `P`, i.e. `(P, _)`,
+  1. Remove all `DownwardMessageQueues` of `P`.
+  1. Remove `RelayDispatchQueueSize` of `P`.
+  1. Remove `RelayDispatchQueues` of `P`.
+  - Note that we don't remove the open/close requests since they are gon die out naturally.
+TODO: What happens with the deposits in channels or open requests?
```

Comment:
great question! lol

Reply:
What happens to DOTs owned by offboarded paras more generally? A parachain with a large locked depsoit for an auction should become a parathread, which should then be able to dispose of it's free DOTs. But parathreads will in general have a small deposit for being a parathread, which means that it will own DOTs when it is offboarded.

So their probably should be an account these DOTs go to, by default the treasury. If these deposits are in the parachain account but just reserved, then we might not need to clean them up specifically here. So what happens to the para's account on the relay chain when it is offboarded?

Reply:
> What happens to DOTs owned by offboarded paras more generally?

The `slots` module maintains an Offboarding address for each parachain. This is initially set to the address of the bidder, but the parachain can submit a `Call` to change it to an address of its choosing. When the parachain is offboarded, its deposit is sent to the given address.